packages/opentelemetry-core/src/context/propagation/JaegerHttpTraceFormat.ts

```diff
@@ -0,0 +1,112 @@
+ * {trace-id}
+ * 64-bit or 128-bit random number in base16 format
+ * Can be variable length, shorter values are 0-padded on the left
+ * Clients in some languages support 128-bit, migration pending
+ * Value of 0 is invalid
+ * {span-id}
+ * 64-bit random number in base16 format
+ * {parent-span-id}
+ * 64-bit value in base16 format representing parent span id
+ * Deprecated, most Jaeger clients ignore on the receiving side, but still include it on the sending side
+ * 0 value is valid and means “root span” (when not ignored)
+ * {flags}
+ * One byte bitmap, as two hex digits
+ * Bit 1 (right-most, least significant) is “sampled” flag
+ * 1 means the trace is sampled and all downstream services are advised to respect that
+ * 0 means the trace is not sampled and all downstream services are advised to respect that
+ * We’re considering a new feature that allows downstream services to upsample if they find their tracing level is too low
+ * Bit 2 is “debug” flag
+ * Debug flag should only be set when the sampled flag is set
+ * Instructs the backend to try really hard not to drop this trace
+ * Other bits are unused.
+ * Inspired by jaeger-client-node project
+ */
+export class JaegerHttpTraceFormat implements HttpTextFormat {
+  inject(
+    spanContext: SpanContext,
+    format: string,
+    carrier: { [key: string]: unknown }
+  ) {
+    const hexTraceId = removeLeadingZeros(spanContext.traceId);
+    const hexSpanId = removeLeadingZeros(spanContext.spanId);
+    const parentSpanId = '0';
+    const flags = TraceFlags.SAMPLED;
+
+    carrier[
+      UBER_TRACE_ID_HEADER
+    ] = `${hexTraceId}:${hexSpanId}:${parentSpanId}:${flags}`;
+  }
+
+  extract(
+    format: string,
+    carrier: { [key: string]: unknown }
+  ): SpanContext | null {
+    const uberTraceIdHeader = carrier[UBER_TRACE_ID_HEADER];
+    if (!uberTraceIdHeader) return null;
+    const uberTraceId = Array.isArray(uberTraceIdHeader)
+      ? uberTraceIdHeader[0]
+      : uberTraceIdHeader;
+
+    return deserializeSpanContext(uberTraceId);
+  }
+}
+
+/**
+ * @param {string} input - the input for which leading zeros should be removed.
+ * @return {string} - returns the input string without leading zeros.
+ **/
+function removeLeadingZeros(input: string): string {
+  let counter = 0;
+  let length = input.length - 1;
+  for (let i = 0; i < length; i++) {
+    if (input.charAt(i) === '0') {
+      counter++;
+    } else {
+      break;
+    }
+  }
+
+  return input.substring(counter);
+}
+
+/**
+ * @param {string} serializedString - a serialized span context.
+ * @return {SpanContext} - returns a span context represented by the serializedString.
+ **/
+function deserializeSpanContext(serializedString: string): SpanContext | null {
+  let headers = serializedString.split(':');
+  if (headers.length !== 4) {
+    return null;
+  }
+  const [traceId, spanId, , flags] = headers;
+
+  const traceFlags =
+    Number('0x' + (isNaN(Number(flags)) ? 1 : Number(flags))) & 1;
```

Comment:
There are a lot of number conversions here. Let's take the most common case where `flags` is a valid hex number. The ternary converts it to a number, evaluates to false, and converts it to a number again. Then concatenates it to `'0x'` which coerces it to a string, then parses it with `Number` again.

This is converting a string to a number twice, then back to a string, only to convert it back to a number. It seems hardly the most efficient way to do this.

```typescript
// Using parseInt lets you guarantee the correct radix
const parsedFlags = parseInt(flags, 16); // parse number once, no string conversions
const traceFlags = isNan(parsedFlags) ? 1 : parsedFlags & 1;
```

This seems to me more readable, while also drastically more efficient.